### DISCUSSION TITLE
Dependency injection of SQLAlchemy (1.4.51) sessions causes Deadlock.

### DISCUSSION THREAD

**Discussion by webertim** [fastapi/fastapi#11186] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Iterator
from fastapi import Depends, FastAPI
from sqlalchemy import create_engine
from sqlalchemy import Column, Integer, String
from sqlalchemy.orm import Session
from sqlalchemy.ext.declarative import declarative_base

import uvicorn

app = FastAPI()


Base = declarative_base()

class Item(Base):
    __tablename__ = "items"

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String, index=True)
    description = Column(String, index=True)


db_url = "postgresql://user:pw@localhost:5432/db"
engine = create_engine(db_url, echo=True, connect_args={"options": "-c lock_timeout=5000"})


def get_db() -> Iterator[Session]:
    try:
        session = Session(engine)
        yield session
    finally:
        session.close()

@app.get("/dummy")
def dummy(
    session: Session = Depends(get_db)
) -> str:
    a = session.get(Item, "1")

    if not a:
        return ""

    return f"{a.name, a.description}"


if __name__ == "__main__":
    uvicorn.run(app , host="0.0.0.0", port=8000, reload=False, log_level="debug", workers=1)
```

### Description

I am using FastAPI with SQLAlchemy in my backend. I use dependency injection for creating and closing sessions as can be seen in the example code. Since my whole backend uses sync database operations I rely on simple function definitions (to execute the requests in a thread pool). The code works fine as long as I send a minimal amount of requests in parallel. I also wrote a simple test script for executing many requests in parallel. (https://github.com/tiangolo/fastapi/blob/6336604906d50e0aca978d15ef82236bac23fb15/fastapi/routing.py#L144-L149)

```python
import asyncio
import httpx

async def get_print(client: httpx.AsyncClient):
    res = await client.get("http://localhost:8000/dummy")
    print(res.text)

async def main():
    client = httpx.AsyncClient()
    
    tasks = []
    for _ in range(100):
        tasks.append(get_print(client))
        
    await asyncio.wait(tasks)
    
if __name__ == '__main__':
    asyncio.run(main())
```

Once I execute the script the backend code ends up in a deadlock. One way to 'fix' this, is to use an async function for the endpoints as well as the dependency injection function. But this is not really a fix in my case since it would run all requests sequentially.

I also did some digging and found a similar issue that got fixed a long time ago (https://github.com/tiangolo/fastapi/issues/3205). I can confirm that the code that fixed the issue is included in my local code. I also did some debugging and found that the worker threads never actually return. Even those that successfully ran an endpoint function. 

If I call `session.close()` manually inside of the endpoint everything works again.

Since this behavior is reproducible by such a minimal example and follows the docs for handling database sessions, I feel like this is a bug.

I will continue to investigate and submit a PR with a fix if I find a solution. Nevertheless I would appreciate any help :)

**Edit:** I think I found the reason for the deadlock. The field validation is done in the threadpool if the handler is defined with `def`. If the whole threadpool is already blocked by waiting for a free session but the sessions are only freed after the validation this causes a deadlock. I am not familiar enough with the code to understand why the validation is executed on the threadpool for sync functions only. Once I always run validation synchronously the issue goes away. If someone can confirm that this is ok to do I will happily submit a PR.

**Edit2:** Found the reason for running it in a threadpool https://github.com/tiangolo/fastapi/pull/888. I'll create a PR with another approach.

**Edit3:** One last edit... What do you think is the best approach to handle this? is it ok to simply end the AsyncExitStack before seriailizing the response? Or does it make sense to use a different thread pool for serialization? I think opinions of core maintainers would be very helpful here.

### Operating System

Linux, Windows

### Operating System Details

WSL in Windows

### FastAPI Version

0.109.2

### Pydantic Version

1.10.14

### Python Version

Python 3.10.12

**Comment by YuriiMotov** [comment]
Is this still an issue in latest version?
Let's close this for now to keep the discussions clean and organized.
If you still need help, feel free to let us know.
